File: frontend/services/sessions.ts
import { apiRequest } from "../lib/api";
import type {
  AnalysisStepStatus,
  ChoquetAggregationResult,
  DeterministicFeedbackResult,
  FrameBatchRequest,
  FrameBatchResponse,
  DeterministicEvaluationResult,
  FuzzyInterpretationResult,
  IT2FuzzyInterpretationResult,
  LLMFeedbackResult,
  LiveEndRequest,
  LiveReadinessRequest,
  LiveStartResponse,
  OntologyReasoningResult,
  PedagogicalDecisionResult,
  SessionAnalysisPipelineStatus,
  SessionAnalysisStep,
  SessionAnalysisWarning,
  SessionArtifactsResponse,
  SessionCreateRequest,
  TemporalModelingResult,
  TrainingSession,
  UploadProcessingResponse
} from "../types/sessions";

export function createSession(payload: SessionCreateRequest) {
  return apiRequest<TrainingSession>("/sessions", {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

type SessionAnalysisResultKey =
  | "evaluation"
  | "fuzzy"
  | "it2"
  | "feedback"
  | "pedagogy"
  | "ontology"
  | "choquet"
  | "temporal"
  | "llm";

type SessionAnalysisPipelineStep<TResult = unknown> = {
  id: SessionAnalysisStep;
  label: string;
  service: (sessionId: string) => Promise<TResult>;
  resultKey: SessionAnalysisResultKey;
  required: boolean;
  dependencyNotes?: string;
  validate: (
    result: TResult
  ) => {
    ok: boolean;
    message?: string;
    diagnosticFlags: string[];
  };
};

type SessionAnalysisPipelineStepResults = {
  evaluation: DeterministicEvaluationResult;
  fuzzy: FuzzyInterpretationResult;
  it2: IT2FuzzyInterpretationResult;
  feedback: DeterministicFeedbackResult;
  pedagogy: PedagogicalDecisionResult;
  ontology: OntologyReasoningResult;
  choquet: ChoquetAggregationResult;
  temporal: TemporalModelingResult;
  llm: LLMFeedbackResult;
};

export type SessionAnalysisPipelineOutput =
  Partial<SessionAnalysisPipelineStepResults> & {
    status: SessionAnalysisPipelineStatus;
    completedSteps: string[];
    failedRequiredStep?: string;
    warnings: SessionAnalysisWarning[];
    artifacts?: SessionArtifactsResponse;
  };

type RunSessionAnalysisOptions = {
  onProgress?: (update: {
    stepId: SessionAnalysisStep;
    status: AnalysisStepStatus;
    warning?: SessionAnalysisWarning;
  }) => void;
};

export function getSession(sessionId: string) {
  return apiRequest<TrainingSession>(`/sessions/${sessionId}`);
}

export function getRecentSessions(limit = 10) {
  return apiRequest<TrainingSession[]>(`/sessions/recent?limit=${limit}`);
}

export function getSessionArtifacts(sessionId: string) {
  return apiRequest<SessionArtifactsResponse>(`/sessions/${sessionId}/artifacts`);
}

export function runSessionEvaluation(sessionId: string) {
  return apiRequest<DeterministicEvaluationResult>(`/sessions/${sessionId}/evaluate`, {
    method: "POST"
  });
}

export function generateDeterministicSessionFeedback(sessionId: string) {
  return apiRequest<DeterministicFeedbackResult>(`/sessions/${sessionId}/feedback`, {
    method: "POST"
  });
}

export function runSessionFuzzyInterpretation(sessionId: string) {
  return apiRequest<FuzzyInterpretationResult>(
    `/sessions/${sessionId}/interpret/fuzzy`,
    {
      method: "POST"
    }
  );
}

export function runSessionIT2FuzzyInterpretation(sessionId: string) {
  return apiRequest<IT2FuzzyInterpretationResult>(
    `/sessions/${sessionId}/interpret/it2-fuzzy`,
    {
      method: "POST"
    }
  );
}

export function runSessionPedagogy(sessionId: string) {
  return apiRequest<PedagogicalDecisionResult>(`/sessions/${sessionId}/pedagogy`, {
    method: "POST"
  });
}

export function runSessionOntologyReasoning(sessionId: string) {
  return apiRequest<OntologyReasoningResult>(`/sessions/${sessionId}/ontology`, {
    method: "POST"
  });
}

export function runSessionChoquetAggregation(sessionId: string) {
  return apiRequest<ChoquetAggregationResult>(
    `/sessions/${sessionId}/aggregate/choquet`,
    {
      method: "POST"
    }
  );
}

export function runSessionTemporalModeling(sessionId: string) {
  return apiRequest<TemporalModelingResult>(`/sessions/${sessionId}/model/temporal`, {
    method: "POST"
  });
}

export function generateLLMSessionFeedback(sessionId: string) {
  return apiRequest<LLMFeedbackResult>(`/sessions/${sessionId}/feedback/llm`, {
    method: "POST"
  });
}

const REQUIRED_DEPENDENCY_FAILURE_FLAGS = [
  "MISSING_POSE_SEQUENCE",
  "MISSING_FEEDBACK_RESULT",
  "EVALUATION_NOT_COMPLETED"
];

const OPTIONAL_SUCCESS_STATUSES: Partial<Record<SessionAnalysisStep, string[]>> = {
  fuzzy: ["COMPLETED", "NO_INTERPRETABLE_METRICS", "DISABLED"],
  it2: ["COMPLETED", "NO_INTERPRETABLE_METRICS", "DISABLED"],
  pedagogy: ["COMPLETED", "NO_ACTIONABLE_FEEDBACK"],
  ontology: ["COMPLETED", "NO_SIGNIFICANT_ISSUES"],
  choquet: ["COMPLETED", "NO_ACTIONABLE_ISSUES"],
  temporal: ["COMPLETED"],
  llm: ["COMPLETED"]
};

function getErrorText(error: unknown) {
  return error instanceof Error ? error.message : "Unexpected analysis error.";
}

function getStatus(result: unknown) {
  if (result && typeof result === "object" && "status" in result) {
    const status = (result as { status?: unknown }).status;
    return typeof status === "string" ? status : null;
  }

  return null;
}

function getDiagnosticFlags(result: unknown) {
  if (!result || typeof result !== "object") {
    return [];
  }

  const resultWithFlags = result as {
    diagnostic_flags?: unknown;
    context_diagnostic_flags?: unknown;
    dominant_side_diagnostic_flags?: unknown;
  };

  const flags = [
    ...(Array.isArray(resultWithFlags.diagnostic_flags)
      ? resultWithFlags.diagnostic_flags
      : []),
    ...(Array.isArray(resultWithFlags.context_diagnostic_flags)
      ? resultWithFlags.context_diagnostic_flags
      : []),
    ...(Array.isArray(resultWithFlags.dominant_side_diagnostic_flags)
      ? resultWithFlags.dominant_side_diagnostic_flags
      : [])
  ].filter((flag): flag is string => typeof flag === "string" && flag.length > 0);

  return Array.from(new Set(flags));
}

function hasRequiredDependencyFailure(diagnosticFlags: string[]) {
  return diagnosticFlags.some(
    (flag) =>
      REQUIRED_DEPENDENCY_FAILURE_FLAGS.includes(flag) ||
      flag === "EVALUATION_STATUS:FAILED" ||
      flag.startsWith("EVALUATION_STATUS:FAILED")
  );
}

function buildStepWarning(
  step: SessionAnalysisStep | string,
  message: string,
  diagnosticFlags: string[] = []
) : SessionAnalysisWarning {
  return {
    step,
    message,
    diagnosticFlags: Array.from(new Set(diagnosticFlags))
  };
}

function addWarning(
  warnings: SessionAnalysisWarning[],
  warning: SessionAnalysisWarning
) {
  const alreadyRecorded = warnings.some(
    (existingWarning) =>
      existingWarning.step === warning.step &&
      existingWarning.message === warning.message &&
      existingWarning.diagnosticFlags.join("|") === warning.diagnosticFlags.join("|")
  );

  if (!alreadyRecorded) {
    warnings.push(warning);
  }
}

function validateEvaluationResult(result: DeterministicEvaluationResult) {
  const diagnosticFlags = getDiagnosticFlags(result);

  if (hasRequiredDependencyFailure(diagnosticFlags)) {
    return {
      ok: false,
      message: "Analysis could not be completed because evaluation dependencies were missing.",
      diagnosticFlags
    };
  }

  if (result.status !== "COMPLETED") {
    return {
      ok: false,
      message: `Analysis could not be completed because evaluation ended with status ${result.status}.`,
      diagnosticFlags
    };
  }

  return { ok: true, diagnosticFlags };
}

function validateDeterministicFeedbackResult(result: DeterministicFeedbackResult) {
  const diagnosticFlags = getDiagnosticFlags(result);

  if (hasRequiredDependencyFailure(diagnosticFlags)) {
    return {
      ok: false,
      message: "Analysis could not be completed because feedback dependencies were missing.",
      diagnosticFlags
    };
  }

  if (result.status !== "COMPLETED" && result.status !== "NO_ACTIONABLE_ISSUES") {
    return {
      ok: false,
      message: `Analysis could not be completed because feedback ended with status ${result.status}.`,
      diagnosticFlags
    };
  }

  return { ok: true, diagnosticFlags };
}

function validateOptionalResult(
  stepId: SessionAnalysisStep,
  result: unknown
) {
  const status = getStatus(result);
  const diagnosticFlags = getDiagnosticFlags(result);

  if (hasRequiredDependencyFailure(diagnosticFlags)) {
    return {
      ok: false,
      message: `${stepId} could not run because a required upstream artifact was missing.`,
      diagnosticFlags
    };
  }

  if (
    stepId === "llm" &&
    result &&
    typeof result === "object" &&
    "fallback_used" in result &&
    (result as { fallback_used?: unknown }).fallback_used === true
  ) {
    return {
      ok: false,
      message: "LLM enhancement was unavailable; deterministic coaching feedback is being shown.",
      diagnosticFlags
    };
  }

  const successStatuses = OPTIONAL_SUCCESS_STATUSES[stepId] ?? ["COMPLETED"];

  if (status && !successStatuses.includes(status)) {
    return {
      ok: false,
      message: `${stepId} ended with status ${status}.`,
      diagnosticFlags
    };
  }

  return { ok: true, diagnosticFlags };
}

function validateFinalArtifacts(artifacts: SessionArtifactsResponse) {
  if (!artifacts.evaluation_result) {
    return buildStepWarning(
      "evaluation",
      "Analysis could not be completed because the evaluation artifact is missing.",
      ["MISSING_EVALUATION_RESULT"]
    );
  }

  const evaluationValidation = validateEvaluationResult(artifacts.evaluation_result);
  if (!evaluationValidation.ok) {
    return buildStepWarning(
      "evaluation",
      evaluationValidation.message ?? "Analysis could not be completed because evaluation failed.",
      evaluationValidation.diagnosticFlags
    );
  }

  if (!artifacts.feedback_result) {
    return buildStepWarning(
      "deterministic_feedback",
      "Analysis could not be completed because deterministic feedback is missing.",
      ["MISSING_FEEDBACK_RESULT"]
    );
  }

  const feedbackValidation = validateDeterministicFeedbackResult(
    artifacts.feedback_result
  );
  if (!feedbackValidation.ok) {
    return buildStepWarning(
      "deterministic_feedback",
      feedbackValidation.message ??
        "Analysis could not be completed because deterministic feedback failed.",
      feedbackValidation.diagnosticFlags
    );
  }

  return null;
}

export const SESSION_ANALYSIS_PIPELINE: ReadonlyArray<SessionAnalysisPipelineStep> = [
  {
    id: "evaluation",
    label: "Evaluating movement",
    service: runSessionEvaluation,
    resultKey: "evaluation",
    required: true,
    dependencyNotes: "Requires a pose_sequence artifact.",
    validate: (result) =>
      validateEvaluationResult(result as DeterministicEvaluationResult)
  },
  {
    id: "fuzzy",
    label: "Interpreting performance",
    service: runSessionFuzzyInterpretation,
    resultKey: "fuzzy",
    required: false,
    dependencyNotes: "Requires completed evaluation.",
    validate: (result) => validateOptionalResult("fuzzy", result)
  },
  {
    id: "it2",
    label: "Refining uncertainty",
    service: runSessionIT2FuzzyInterpretation,
    resultKey: "it2",
    required: false,
    dependencyNotes: "Requires completed evaluation and fuzzy-compatible metrics.",
    validate: (result) => validateOptionalResult("it2", result)
  },
  {
    id: "deterministic_feedback",
    label: "Building coaching feedback",
    service: generateDeterministicSessionFeedback,
    resultKey: "feedback",
    required: true,
    dependencyNotes: "Requires completed deterministic evaluation.",
    validate: (result) =>
      validateDeterministicFeedbackResult(result as DeterministicFeedbackResult)
  },
  {
    id: "pedagogy",
    label: "Selecting coaching focus",
    service: runSessionPedagogy,
    resultKey: "pedagogy",
    required: false,
    dependencyNotes: "Requires deterministic feedback before coaching focus selection.",
    validate: (result) => validateOptionalResult("pedagogy", result)
  },
  {
    id: "ontology",
    label: "Mapping movement concepts",
    service: runSessionOntologyReasoning,
    resultKey: "ontology",
    required: false,
    dependencyNotes: "Requires deterministic feedback and evaluation context.",
    validate: (result) => validateOptionalResult("ontology", result)
  },
  {
    id: "choquet",
    label: "Combining issue patterns",
    service: runSessionChoquetAggregation,
    resultKey: "choquet",
    required: false,
    dependencyNotes: "Requires ontology reasoning and deterministic feedback.",
    validate: (result) => validateOptionalResult("choquet", result)
  },
  {
    id: "temporal",
    label: "Reviewing timing",
    service: runSessionTemporalModeling,
    resultKey: "temporal",
    required: false,
    dependencyNotes: "Requires pose sequence and evaluation phase ranges.",
    validate: (result) => validateOptionalResult("temporal", result)
  },
  {
    id: "llm",
    label: "Preparing coaching summary",
    service: generateLLMSessionFeedback,
    resultKey: "llm",
    required: false,
    dependencyNotes: "Runs last and falls back to deterministic feedback when unavailable.",
    validate: (result) => validateOptionalResult("llm", result)
  }
];

export async function runSessionAnalysisPipeline(
  sessionId: string,
  options: RunSessionAnalysisOptions = {}
) : Promise<SessionAnalysisPipelineOutput> {
  const { onProgress } = options;
  const results: Partial<SessionAnalysisPipelineStepResults> = {};
  const completedSteps: string[] = [];
  const warnings: SessionAnalysisWarning[] = [];

  async function buildFailedResult(
    failedRequiredStep: string,
    warning: SessionAnalysisWarning
  ) : Promise<SessionAnalysisPipelineOutput> {
    addWarning(warnings, warning);

    let artifacts: SessionArtifactsResponse | undefined;
    try {
      artifacts = await getSessionArtifacts(sessionId);
    } catch {
      artifacts = undefined;
    }

    return {
      ...results,
      status: "FAILED",
      completedSteps,
      failedRequiredStep,
      warnings,
      artifacts
    };
  }

  for (const step of SESSION_ANALYSIS_PIPELINE) {
    onProgress?.({ stepId: step.id, status: "RUNNING" });

    try {
      const result = await step.service(sessionId);
      (results as Record<SessionAnalysisResultKey, unknown>)[step.resultKey] = result;
      const validation = step.validate(result);

      if (!validation.ok) {
        const warning = buildStepWarning(
          step.id,
          validation.message ?? `${step.label} did not complete successfully.`,
          validation.diagnosticFlags
        );

        if (step.required) {
          onProgress?.({ stepId: step.id, status: "FAILED", warning });
          return buildFailedResult(step.id, warning);
        }

        addWarning(warnings, warning);
        onProgress?.({ stepId: step.id, status: "WARNING", warning });
        continue;
      }

      completedSteps.push(step.id);
      onProgress?.({ stepId: step.id, status: "COMPLETED" });
    } catch (error) {
      const warning = buildStepWarning(
        step.id,
        `${step.label} failed: ${getErrorText(error)}`,
        []
      );

      if (step.required) {
        onProgress?.({ stepId: step.id, status: "FAILED", warning });
        return buildFailedResult(step.id, warning);
      }

      addWarning(warnings, warning);
      onProgress?.({ stepId: step.id, status: "WARNING", warning });
    }
  }

  let artifacts: SessionArtifactsResponse;
  try {
    artifacts = await getSessionArtifacts(sessionId);
  } catch (error) {
    const warning = buildStepWarning(
      "artifacts",
      `Analysis artifacts could not be loaded: ${getErrorText(error)}`,
      []
    );
    addWarning(warnings, warning);

    return {
      ...results,
      status: "FAILED",
      completedSteps,
      failedRequiredStep: "artifacts",
      warnings
    };
  }

  const artifactFailure = validateFinalArtifacts(artifacts);
  if (artifactFailure) {
    addWarning(warnings, artifactFailure);

    return {
      ...results,
      status: "FAILED",
      completedSteps,
      failedRequiredStep: artifactFailure.step,
      warnings,
      artifacts
    };
  }

  return {
    ...results,
    status: warnings.length ? "COMPLETED_WITH_WARNINGS" : "COMPLETED",
    completedSteps,
    warnings,
    artifacts
  };
}

export function uploadSessionVideo(sessionId: string, file: File) {
  const formData = new FormData();
  formData.append("file", file);

  return apiRequest<UploadProcessingResponse>(`/sessions/${sessionId}/upload`, {
    method: "POST",
    body: formData
  });
}

export function startLiveSession(
  sessionId: string,
  payload: LiveReadinessRequest
) {
  return apiRequest<LiveStartResponse>(`/sessions/${sessionId}/live/start`, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export function submitLiveFrameBatch(
  sessionId: string,
  payload: FrameBatchRequest
) {
  return apiRequest<FrameBatchResponse>(`/sessions/${sessionId}/live/frame-batch`, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export function endLiveSession(sessionId: string, payload: LiveEndRequest) {
  return apiRequest<TrainingSession>(`/sessions/${sessionId}/live/end`, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export const evaluateSession = runSessionEvaluation;
export const generateSessionFeedback = generateDeterministicSessionFeedback;
export const generateFuzzySessionInterpretation = runSessionFuzzyInterpretation;
export const submitSessionUpload = uploadSessionVideo;

File: backend/app/api/routes/sessions.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, File, Query, UploadFile, status

from app.core.dependencies import get_current_user, get_session_service
from app.models.user import User
from app.schemas.session import (
    ChoquetAggregationResult,
    DeterministicEvaluationResult,
    DeterministicFeedbackResult,
    FrameBatchRequest,
    FrameBatchResponse,
    FuzzyInterpretationResult,
    IT2FuzzyInterpretationResult,
    LLMFeedbackResult,
    LiveEndRequest,
    LiveReadinessRequest,
    LiveStartResponse,
    OntologyReasoningResult,
    PedagogicalDecisionResult,
    SessionCreateRequest,
    SessionArtifactsResponse,
    SessionResponse,
    UploadProcessingResponse,
)
from app.services.session_service import SessionService

router = APIRouter(prefix="/sessions", tags=["sessions"])


@router.post("", response_model=SessionResponse, status_code=status.HTTP_201_CREATED)
def create_session(
    payload: SessionCreateRequest,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> SessionResponse:
    return session_service.create_session(user_id=current_user.id, payload=payload)


@router.get("/recent", response_model=list[SessionResponse])
def list_recent_sessions(
    limit: int = Query(default=10, ge=1, le=25),
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> list[SessionResponse]:
    return session_service.list_recent_sessions(user_id=current_user.id, limit=limit)


@router.get("/{session_id}", response_model=SessionResponse)
def get_session(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> SessionResponse:
    return session_service.get_session(user_id=current_user.id, session_id=session_id)


@router.get(
    "/{session_id}/artifacts",
    response_model=SessionArtifactsResponse,
)
def get_session_artifacts(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> SessionArtifactsResponse:
    return session_service.get_session_artifacts(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/evaluate",
    response_model=DeterministicEvaluationResult,
)
def evaluate_session(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> DeterministicEvaluationResult:
    return session_service.evaluate_session(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/feedback",
    response_model=DeterministicFeedbackResult,
)
def generate_session_feedback(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> DeterministicFeedbackResult:
    return session_service.generate_feedback(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/interpret/fuzzy",
    response_model=FuzzyInterpretationResult,
)
def generate_fuzzy_session_interpretation(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> FuzzyInterpretationResult:
    return session_service.generate_fuzzy_interpretation(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/interpret/it2-fuzzy",
    response_model=IT2FuzzyInterpretationResult,
)
def generate_it2_fuzzy_session_interpretation(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> IT2FuzzyInterpretationResult:
    return session_service.generate_it2_fuzzy_interpretation(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/pedagogy",
    response_model=PedagogicalDecisionResult,
)
def generate_session_pedagogy(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> PedagogicalDecisionResult:
    return session_service.generate_pedagogical_decision(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/ontology",
    response_model=OntologyReasoningResult,
)
def generate_session_ontology_reasoning(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> OntologyReasoningResult:
    return session_service.generate_ontology_reasoning(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/aggregate/choquet",
    response_model=ChoquetAggregationResult,
)
def generate_session_choquet_aggregation(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> ChoquetAggregationResult:
    return session_service.generate_choquet_aggregation(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/feedback/llm",
    response_model=LLMFeedbackResult,
)
def generate_llm_session_feedback(
    session_id: UUID,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> LLMFeedbackResult:
    return session_service.generate_llm_feedback(
        user_id=current_user.id,
        session_id=session_id,
    )


@router.post(
    "/{session_id}/upload",
    response_model=UploadProcessingResponse,
    response_model_exclude_none=True,
)
def upload_session_media(
    session_id: UUID,
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> UploadProcessingResponse:
    file_bytes = file.file.read()
    return session_service.process_upload(
        user_id=current_user.id,
        session_id=session_id,
        file_name=file.filename,
        content_type=file.content_type,
        file_size_bytes=len(file_bytes),
        file_bytes=file_bytes,
    )


@router.post("/{session_id}/live/start", response_model=LiveStartResponse)
def start_live_session(
    session_id: UUID,
    payload: LiveReadinessRequest,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> LiveStartResponse:
    return session_service.start_live_session(
        user_id=current_user.id,
        session_id=session_id,
        payload=payload,
    )


@router.post("/{session_id}/live/frame-batch", response_model=FrameBatchResponse)
def submit_frame_batch(
    session_id: UUID,
    payload: FrameBatchRequest,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> FrameBatchResponse:
    return session_service.accept_live_frame_batch(
        user_id=current_user.id,
        session_id=session_id,
        payload=payload,
    )


@router.post("/{session_id}/live/end", response_model=SessionResponse)
def end_live_session(
    session_id: UUID,
    payload: LiveEndRequest,
    current_user: User = Depends(get_current_user),
    session_service: SessionService = Depends(get_session_service),
) -> SessionResponse:
    return session_service.end_live_session(
        user_id=current_user.id,
        session_id=session_id,
        payload=payload,
    )
